### DISCUSSION TITLE
How to configure CSS path absolute?

### DISCUSSION THREAD

**Discussion by Stwissel** [mkdocs/mkdocs#3153] [Q&A]
Currently CSS and JS links from templates seem to be rendered relative to a page:
assets/some.css for index.html
../assets/some.css for subdir/page.css

Works nicely, except for a custom 404 page that can be rendered at ant path.

Is there a way to tell mkdocs to render css/js path absolute starting with / ?

**Comment by Andre601** [comment]
Not an expert with templating, but I'm sure you can use something like `{{ base_url }}/assets/some.css` and it would always use the base url defined in the mkdocs.yml file

**Reply by Stwissel** [reply]
Unfortunately it isn't that simple. When We look at the stle section of `base.html` we find:

```html
      {%- block styles %}
        <link href="{{ 'css/bootstrap.min.css'|url }}" rel="stylesheet">
        <link href="{{ 'css/font-awesome.min.css'|url }}" rel="stylesheet">
        <link href="{{ 'css/base.css'|url }}" rel="stylesheet">
        {%- if config.theme.highlightjs %}
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.5.0/styles/{{ config.theme.hljs_style }}.min.css">
        {%- endif %}
        {%- for path in extra_css %}
        <link href="{{ path }}" rel="stylesheet">
        {%- endfor %}
      {%- endblock %}
```

Which translates in my case to:

```html
<link rel="stylesheet" href="assets/stylesheets/main.7bf56d0a.min.css">
<link rel="stylesheet" href="assets/stylesheets/palette.a0c5b2b5.min.css">
<link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
<link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
<style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
<!-- THIS ONE I CAN CONTROL -->
<link rel="stylesheet" href="/Domino-rest-api/assets/css/keep_style.css">
```

So mkdocs packages the stylesheet which is nice, but I don't see (due to lack of knowledge probably) how I could turn `assets/stylesheets/main.7bf56d0a.min.css` into `/Domino-rest-api/assets/stylesheets/main.7bf56d0a.min.css` with `/Domino-rest-api` being my site prefix in `mkdocs.yml`

Help is appreciated

**Reply by kamilkrzyskow** [reply]
@Stwissel The `base.html` example doesn't align with what is shown in the result. The `base.html` looks like this: 
https://github.com/mkdocs/mkdocs/blob/master/mkdocs/themes/mkdocs/base.html#L19-L29 (from the "mkdocs" theme)
and result looks looks more like this:
https://github.com/squidfunk/mkdocs-material/blob/master/material/base.html#L42-L65 (from the "Material" theme, post-build)
https://github.com/squidfunk/mkdocs-material/blob/master/src/base.html#L85-L131 (from the "Material" theme, pre-build)

However, the Material theme joins the custom paths with the base url and the mkdocs theme doesn't 🤔, So I guess your comment 
`<!-- THIS ONE I CAN CONTROL -->` indicates you did something like this in your `mkdocs.yml`:
```yml
extra_css:
  - Domino-rest-api/assets/css/keep_style.css
```
and then the Material theme joined it with the `/` and et voila it's `/Domino-rest-api/assets/css/keep_style.css`

---

> with /Domino-rest-api being my site prefix in mkdocs.yml

What is the setting you called "site prefix" here?

---

The solution is probably to use the [`site_url`](https://www.mkdocs.org/user-guide/configuration/#site_url) setting in `mkdocs.yml`:
Make sure that you have `site_url: https://example.com/Domino-rest-api/` set.
Then in your `extra_css` use `assets/css/keep_style.css`

**Reply by Stwissel** [reply]
Sorry for not being clear. It isn't about the `extra_css`, it is about the theme generated CSS.
I'm using material. Reproducer [is here](https://stwissel.github.io/mknotfound/)

In my `mkdocs.yml` I think I have everything (prefix is actually the site_url, sorry for confusing terms)"

```yaml
site_name: Trial with 404
site_url: https://stwissel.github.io/mknotfound/
use_directory_urls: false
theme:
   name: material
   language: en
   features:
     - navigation.tabs
     - navigation.indexes
     - navigation.top

# Repository
repo_url: https://github.com/Stwissel/mknotfound
repo_name: mknotfound
edit_uri: edit/main/docs
```

The generated page contains link to the Theme CSS:

```html
<link rel="stylesheet" href="assets/stylesheets/palette.a0c5b2b5.min.css">
<link rel="stylesheet" href="assets/stylesheets/main.c4a75a56.min.css">
```

This is all fine and good until you want a customised 4040 page. On the [first level](https://stwissel.github.io/mknotfound/nothere) is looks fine, but breaks when the invalid URL [contains a path](https://stwissel.github.io/mknotfound/not/here). There is no custom CSS involved.

Both mkdocs and mkdocs material render their theme CSS relative. What I wonder: how can they be configured to render these CSS (and probably js (?)) path absolute (like you can do with `extra_css` Sorry again for not being clear

**Reply by kamilkrzyskow** [reply]
OK, I've finally recalled that you had this issue on the Material discussion forums, and made an Issue in this repository later. I've found the issue https://github.com/mkdocs/mkdocs/issues/3140 and commented my first thoughts about the matter at hand. Once I'll figure something out I'll post it here.

**Comment by kamilkrzyskow** [comment]
**UPDATE:**
I've finished the hook, and now it has working absolute links in every regard in that minimal reproduction. Just like I thought, it's possible to modify the context for a single page. However checking such condition if it's the 404 page might have some noticeable performance impact at 1-5k+ docs pages.

```py
def on_page_context(context: Dict[str, Any], *, page: Page, **__) -> Optional[Dict[str, Any]]:
    """Runs before each page is built and joined with the theme."""
    # LOG.debug(f'{HOOK_NAME}: Running "on_page_context"')

    # Only process 404.html page
    if page.file.dest_path.endswith("404.html"):
        # Change the base_url for the search_index.json file
        context["base_url"] = BASE_URL

        # Change every url to another page to be absolute
        for p in context["pages"]:
            p: File
            p.url = relative_to_absolute(p.url)

    return context
```

[working_404_absolute_paths.zip](https://github.com/mkdocs/mkdocs/files/11001307/working_404_absolute_paths.zip)


---

So I've worked for some time and managed to get the absolute `assets/` paths working. Unfortunately creating the 404 page as part of the content creates multiple other path issues, like every navigation link is relative too, and the search_index.json is relative too. 
Perhaps it's possible to modify the context of a single page, and modify the `base_url` for that specific page which would maybe help, but also might not. Getting this to work and not bug out in would be rather time consuming, I don't think that, I'll be the one to finish it.

My partial solution entails, using a hook which modifies the behavior of the `url` Jinja2 filter used in most links in a template `{{ path | url }}`.
```py
def on_env(env: Environment, *, config: MkDocsConfig, **__) -> Optional[Environment]:
    """Main function. Triggers just before the build begins."""

    LOG.debug(f'{HOOK_NAME}: Running "on_env"')

    global BASE_URL
    BASE_URL = urlsplit(config.site_url or "/").path

    env.filters["url"] = new_filter

    return None


@contextfilter
def new_filter(context, value: str) -> str:
    """A Template filter to normalize URLs (modified to add the absolute path prefix)."""

    result = default_filter(context, value)

    if "assets/" not in result:
        return result

    if not result.startswith("."):
        return f"{BASE_URL}{result}"

    if result.startswith("./"):
        result = result.replace("./", "", 1)

    if result.startswith("../"):
        result = result.replace("../", "")

    return f"{BASE_URL}{result}"
```

--- 

[~~working_CSS_in_404.zip~~](https://github.com/mkdocs/mkdocs/files/10993605/working_CSS_in_404.zip)

**Reply by kamilkrzyskow** [reply]
@Stwissel pinging after the update, check out the finial version if it solves your issue.

**Reply by kamilkrzyskow** [reply]
@Stwissel another ping asking for feedback ✌️ If your issue got resolved then please mark it as the answer.
